sim/src/main/scala/endpoints/TracerVWidget.scala

```diff
@@ -18,43 +18,79 @@ import icenet.IceNIC._
 import junctions.{NastiIO, NastiKey}
 import TokenQueueConsts._
 
-class TraceOutputTop(val numTraces: Int)(implicit val p: Parameters) extends Bundle {
-  val traces = Vec(numTraces, new TracedInstruction)
+// Hack: In a457f658a, RC added the Clocked trait to TracedInstruction, which breaks midas
+// I/O token handling. The non-Clock fields of this Bundle should be factored out in rocket chip.
+// For now, we create second Bundle with Clock (of type Clock) and Reset removed
+class DeclockedTracedInstruction(private val proto: TracedInstruction) extends Bundle {
+  val valid = Bool()
+  val iaddr = UInt(proto.iaddr.getWidth.W)
+  val insn  = UInt(proto.insn.getWidth.W)
+  val priv  = UInt(proto.priv.getWidth.W)
+  val exception = Bool()
+  val interrupt = Bool()
+  val cause = UInt(proto.cause.getWidth.W)
+  val tval = UInt(proto.tval.getWidth.W)
 }
 
-class SimTracerV extends Endpoint {
+object DeclockedTracedInstruction {
+  // Generates a hardware Vec of declockedInsns
+  def fromVec(clockedVec: Vec[TracedInstruction]): Vec[DeclockedTracedInstruction] = {
+    val declockedVec = clockedVec.map(insn => Wire(new DeclockedTracedInstruction(insn.cloneType)))
+    declockedVec.zip(clockedVec).foreach({ case (declocked, clocked) =>
+      declocked.valid := clocked.valid
+      declocked.iaddr := clocked.iaddr
+      declocked.insn := clocked.insn
+      declocked.priv := clocked.priv
+      declocked.exception := clocked.exception
+      declocked.interrupt := clocked.interrupt
+      declocked.cause := clocked.cause
+      declocked.tval := clocked.tval
+    })
+    VecInit(declockedVec)
+  }
 
-  // this is questionable ...
-  // but I can't see a better way to do this for now. getting sharedMemoryTLEdge is the problem.
-  var tracer_param = Parameters.empty
-  var num_traces = 0
+  // Generates a Chisel type from that returned by a Diplomatic node's in() or .out() methods
+  def fromNode(ports: Seq[(Vec[TracedInstruction], Any)]): Seq[Vec[DeclockedTracedInstruction]] = ports.map({
+    case (bundle, _) => Vec(bundle.length, new DeclockedTracedInstruction(bundle.head.cloneType))
+  })
+}
+
+// The IO matched on by the TracerV endpoint: a wrapper around a heterogenous
+// bag of vectors. Each entry is Vec of committed instructions
+class TraceOutputTop(private val traceProto: Seq[Vec[DeclockedTracedInstruction]]) extends Bundle {
+  val traces = HeterogeneousBag(traceProto.map(_.cloneType))
```

Comment:
Why is this a HeterogeneousBag?

Reply:
Different cores might have different traced instructions.